baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+            )
+
+        # Increment iteration counter
+        self.iter += 1
+
+    def writeFullVariableHistory(self, name: str, values: Iterable) -> None:
+        """Write the entire history of a variable in one go
+
+        This function should be used in the case where your solver already handles the recording of variables during a
+        solution (e.g ADflow) but you want to use a SolverHistory object to facilitate writing it to a file
+
+        Parameters
+        ----------
+        name : str
+            Variable name
+        values : Iterable
+            Values to record, will be converted to a list
+        """
+        if name not in self.variables:
+            raise ValueError(
+                f"Unknown variables {name} supplied to Solution History recorder, recorded variables are {self.getVariables()}"
+            )
+        try:
+            self.data[name] = [None if v is None else self.variables[name]["type"](v) for v in values]
+        except ValueError as e:
+            raise TypeError(
+                f"A value provided for variable '{name}' could not be converted to the type declared for this variable: {self.variables[name]['type']}"
+            ) from e
+
+    def printHeader(self) -> None:
+        """Print the header of the iteration printout
+
+        The header will look something like this:
+
+        .. code-block:: text
+
+            +--------------------------------------------------------------------------...------+
+            | Iter  |       Time        |       Var 1       |       Var 2       |      ...      |
+            +--------------------------------------------------------------------------...------+
+
+        """
+
+        # Each field will be `columnWidth` characters wide plus 2 spaces each side, plus the vertical bar between each
+        # field
+        headerString = "|"
+        for var in self.variablesToPrint:
+            headerString += self.variables[var]["headerFormat"].format(var)
+            headerString += "|"
+        headerWidth = len(headerString)
+
+        headerBar = "+" + "-" * (headerWidth - 2) + "+"
+        print(headerBar)
+        print(headerString)
+        print(headerBar)
+
+    def printData(self, iters: Optional[Union[int, Iterable[int]]] = None) -> None:
+        """Print a selection of lines from the history
+
+        Parameters
+        ----------
+        iters : int or Iterable of ints, optional
+            Iteration numbers to print, by default only the last iteration will be printed
+        """
+        if iters is None:
+            iters = [-1]
+        elif isinstance(iters, int):
+            iters = [iters]
+
+        if max(iters) >= self.iter:
+            raise ValueError(
+                f"Requested iteration {max(iters)} is not in the history, only {self.iter} iterations in history"
+            )
+        if min(iters) < -self.iter:
```

Comment:
This will be triggered if the user uses a negative index greater than the length of the history, e.g they want to print iteration -10 but the history is only 5 entries long